### ISSUE TITLE
Email Failure Notification

### ISSUE THREAD

**Issue by m-prodan** [bcgov/foi-flow#2388]
* As an IAO user
* I want to know if the email sent to an applicant has failed
* so that I can investigate and remediate the issue as needed

**Assumptions & Scope**	
This is related to story #2322 and related stories where correspondence is sent to an applicant via email. 

Assumption is that successful emails will be saved to the attachment log, confirming that the email was sent successfully, but we need some kind of notification or flag when an email has NOT been sent successfully (ie: a bounce-back is received)

Will use NR common service, but unsure if it has exception handling. If not will use SMTP server to return error messages for bounceback/undeliverable emails, invalid emails, and not use the common service at all. We need proof that we attempted to send the email.

What is IN scope?
Notification and system generated comment when an email fails
Attachment of bounce back email to the attachment log

What is NOT in scope?
Sending an email

**Acceptance Criteria**	

**Scenario 1: Notification to IAO User**
* GIVEN an email has been sent to an applicant
* WHEN the email fails to be sent for any reason
* THEN a notification will be sent to the assigned IAO analyst with the message "Fee Estimate correspondence failed to send to applicant - see attachment log for details"

**Scenario 2: System Generated Comment**
* GIVEN an email has been sent to an applicant
* WHEN the email fails to be sent for any reason
* THEN a system generated comment will be posted to the comment log of the request with the message "Correspondence failed to send to applicant - see attachment log for details"

**Scenario 3: Bounce-Back Email to Attachment Log**
* GIVEN an email has been sent to an applicant
* WHEN the email fails to be sent for any reason
* THEN the bounce back email will be systematically posted to the attachment log as an EML file
* AND only an IAO user will be able to interact with the attachment

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by m-prodan** [comment]
Updated AC per discussion - please review @lmullane @KaraBeach @arielleandrews @JHarrietha-AOT 

Note that I created two separate modal messages depending on contact information present. We could standardize this, but I thought options should be provided either way.

**Comment by lmullane** [comment]
Hi @m-prodan, @richard-aot, when will receive the notice that the email failed? 

If the notice is immediate and appears as an error message in the modal, then I think the state is not changed (i.e., no reversion)/Instead the user would be presented with the option to cancel to the state change to 'On Hold',' keeping the request in 'Fee Estimate, or 'Save Change' and complete the  the state change to 'On Hold.'

Essentially the state change is not complete until the user saves the change in the modal, which is consistent with our other state change modal behaviour.

**Comment by richard-aot** [comment]
@lmullane , good idea. Yes, we will have response in the modal (it will take a couple seconds). Basically, there will be a backend workflow to send email and to retry a few times if failed, then send response to modal. 

However, as mentioned in refinement meeting, we need to do more research on common service. And see if it handle bounce back and invalid email exception. If not, we will use SMTP server, which supports immediate response for sure.

**Comment by m-prodan** [comment]
Thanks both - updated AC to handle this, assumption will be error message will be more immediate and the error will be noted in the modal as well. I'll keep notifications and system generated comments as this will be important for watchers/request history purposes

**Comment by arielleandrews** [comment]
Decision at sprint planning to have bounce back added to attachments as an email - this will allow the ability to resend the email to the applicant if necessary

**Comment by sumathi-thirumani-aot** [comment]
@arielleandrews @JHarrietha-AOT "Correspondence failed to send to applicant - see attachment log for details" Is this message sufficient to understand that the correspondence is related to fee estimate ? Please review and provide the feedback.

**Comment by m-prodan** [comment]
Updated

**Comment by sumathi-thirumani-aot** [comment]
"541": "Message rejected by the recipient address"
"550 5.1.1": "The email account that you tried to reach does not exist. Please try double-checking the recipient's email address for typos or unnecessary spaces"
"550": "Requested command failed because the user’s mailbox was unavailable, or the receiving server rejected the message because it was likely spam"
"551": "Intended recipient mailbox isn't available on the receiving server"
"552": "Message wasn't sent because the recipient mailbox doesn't have enough storage"
"OTHERS": "Server could not complete the action"

**Comment by arielleandrews** [comment]
@lmullane @m-prodan please review the reasons posted by Sumathi. Thanks!

**Comment by m-prodan** [comment]
These look fine to me - 551 might be difficult to understand, but at the same time it denotes there was an error. Analyst can try again or follow up with applicant to confirm a different or corrected email address.

**Comment by arielleandrews** [comment]
@sumathi-thirumani-aot and @divyav-aot Summarizing my testing so far:
- A valid but non-existent email address does not produce a bounce-back. I used "Arielle.Andrews1@gov.bc.ca"
- An invalid email address did produce a bounce-back. I used "ArielleAndrews123@gov.bc" on request ID JER-2022-87654 and all 3 scenarios passed
- A notification is being sent to the assigned Ministry User - this should be removed, only IAO needs to action a failed email
- May need to do a screenshare with someone from the development team to verify the other error messages work as we may not be able to test those conditions i.e. email inbox is full
